### ISSUE TITLE
Nightly test failures with Cuda builds running on device id != 0

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#5713]
**Describe the bug**
Some unit tests are failing in Cuda builds when running with device_id != 0:

**KokkosCore_UnitTest_Cuda1**
```
4: [ RUN      ] cuda.partitioning_by_args
4: /ascldap/users/ndellin/kokkos/core/unit_test/TestExecSpacePartitioning.hpp:101: Failure
4: Expected equality of these values:
4:   sum1
4:     Which is: 0
4:   N * (N - 1) / 2
4:     Which is: 7642095
4: [  FAILED  ] cuda.partitioning_by_args (182 ms)
4: [ RUN      ] cuda.partitioning_by_vector
4: /ascldap/users/ndellin/kokkos/core/unit_test/TestExecSpacePartitioning.hpp:101: Failure
4: Expected equality of these values:
4:   sum1
4:     Which is: 0
4:   N * (N - 1) / 2
4:     Which is: 7642095
4: [  FAILED  ] cuda.partitioning_by_vector (0 ms)
```

One of the following commits began triggering the failures:

Fixup non-static {Serial,OpenACC}::concurrency() member function unless deprecated code 4 is ON ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/700/changes#052256fa24fd6cb424bef3ab1cd56eb224ef1dff) / [githubweb](https://github.com/kokkos/kokkos/commit/052256fa24fd6cb424bef3ab1cd56eb224ef1dff))
    Make Threads::concurrency() a non-static member function unless deprecated code 4 is ON ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/700/changes#33762bf5aba1c66f25fa8c5803192baf0f74e462) / [githubweb](https://github.com/kokkos/kokkos/commit/33762bf5aba1c66f25fa8c5803192baf0f74e462))
    5594: Remove two-argument CudaLDFFetch constructor ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/700/changes#4e08cb2635df5918b857a3b0c8ef89a611005796) / [githubweb](https://github.com/kokkos/kokkos/commit/4e08cb2635df5918b857a3b0c8ef89a611005796))
    Fixup non-static OpenTarget::concurrency() member function unless deprecated code 4 is ON ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/700/changes#e2efab2a12c766455a57426ae089a147d3d2569a) / [githubweb](https://github.com/kokkos/kokkos/commit/e2efab2a12c766455a57426ae089a147d3d2569a))
    [ci skip] Update changelog ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/700/changes#14e9887310cb3ca5387a8ac92e8515ed2dd39b57) / [githubweb](https://github.com/kokkos/kokkos/commit/14e9887310cb3ca5387a8ac92e8515ed2dd39b57))
    Turn off classic intel workaround in Serial backend ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/700/changes#35fa45c52333e29da5cfb82170d7fac9daa0f279) / [githubweb](https://github.com/kokkos/kokkos/commit/35fa45c52333e29da5cfb82170d7fac9daa0f279))
    Fix classic Intel compiler Serial/OpenMP backend build ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/700/changes#2513fc76770efb3476cb48fc62580c11b586350b) / [githubweb](https://github.com/kokkos/kokkos/commit/2513fc76770efb3476cb48fc62580c11b586350b))
    Add partition_space to OpenMP (#5105) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/700/changes#e4cce6e449d482e2bf0653d3b31c38d95d4090e9) / [githubweb](https://github.com/kokkos/kokkos/commit/e4cce6e449d482e2bf0653d3b31c38d95d4090e9))
Drop CudaLDGFetch special member functions ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#8aa23152c5175dd5ebee91d6856b50c9b1567002) / [githubweb](https://github.com/kokkos/kokkos/commit/8aa23152c5175dd5ebee91d6856b50c9b1567002))
    Drop CudaLDGFetch default constructor definition (prefer defaulted one) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#d9ea15a732762fda6b791547c49b22e3735c00e3) / [githubweb](https://github.com/kokkos/kokkos/commit/d9ea15a732762fda6b791547c49b22e3735c00e3))
    Drop unnecessary inline specifiers in CudaLDGFetch ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#588dda891c01cae8f17671ca00f7bc2eac31246e) / [githubweb](https://github.com/kokkos/kokkos/commit/588dda891c01cae8f17671ca00f7bc2eac31246e))
    Fix {Cuda:: -> Cuda().}concurrency() occurences ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#ad6c3dafec65e6b220a3661fb3ffaf79392c9f1f) / [githubweb](https://github.com/kokkos/kokkos/commit/ad6c3dafec65e6b220a3661fb3ffaf79392c9f1f))
    Replace Impl::CudaInternal::m_maxConcurrency data member by Impl::CudaInternal::concurrency() member function ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#136abca269b6d87de69509c3af59f04eda62093b) / [githubweb](https://github.com/kokkos/kokkos/commit/136abca269b6d87de69509c3af59f04eda62093b))
    Fix chicken and egg Cuda concurrency issue when initializing lock arrays ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#4cfa416c7c4768ab4937419280d6d279f39fea15) / [githubweb](https://github.com/kokkos/kokkos/commit/4cfa416c7c4768ab4937419280d6d279f39fea15))
    Make Cuda::concurrency() member function non static ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#a1b86463689cdd7e798d94794aa0c1568cda2897) / [githubweb](https://github.com/kokkos/kokkos/commit/a1b86463689cdd7e798d94794aa0c1568cda2897))
    Fix {HIP:: -> HIP().}concurrency() occurences ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#965998da585dd96e404dda48474f33df86a1dde4) / [githubweb](https://github.com/kokkos/kokkos/commit/965998da585dd96e404dda48474f33df86a1dde4))
    Introduce Impl::HIPInternal::concurrency static member function ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#969490a83c7930fdb2d8874f93c775a60b01f13e) / [githubweb](https://github.com/kokkos/kokkos/commit/969490a83c7930fdb2d8874f93c775a60b01f13e))
    Make HIP::concurrency() member function non static ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#95fa3a75ebb666056d0b48ea83a3a2082bc53403) / [githubweb](https://github.com/kokkos/kokkos/commit/95fa3a75ebb666056d0b48ea83a3a2082bc53403))
    Fixup HIP header include incomplete type Impl::HIPInternal ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#17170f6eafc7c7c090b77c53ada56e9cd9d56715) / [githubweb](https://github.com/kokkos/kokkos/commit/17170f6eafc7c7c090b77c53ada56e9cd9d56715))
    Add architecture flags for MSVC (#5673) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#f3cf72a2408b56479a09234fe18262bcacb00c82) / [githubweb](https://github.com/kokkos/kokkos/commit/f3cf72a2408b56479a09234fe18262bcacb00c82))
    #5667: dont install std algorithm headers multiple times (#5670) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#8303668181ef68e1684a90d25b460188843249c0) / [githubweb](https://github.com/kokkos/kokkos/commit/8303668181ef68e1684a90d25b460188843249c0))
    Work around CUDA+Clang Thrust issue (#5660) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkosdev2_cuda110_gcc920_cpp17-deviceid_1/701/changes#b2e2096e933e04bf338007696532009f1d467429) / [githubweb](https://github.com/kokkos/kokkos/commit/b2e2096e933e04bf338007696532009f1d467429))

**Please also include the following items to support reproducing the bug**
**Reproducer** (kokkos-dev-2):
```bash
module load sems-archive-env sems-archive-cmake/3.17.1 sems-archive-gcc/9.2.0 cuda/11.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,Serial --arch=Volta70 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="17" --kokkos-path=$KOKKOS_PATH --with-cuda-options=enable_lambda  --no-examples

make -j20

KOKKOS_DEVICE_ID=1 ctest -R KokkosCore_UnitTest_Cuda1 -V

```

**Comment by ajpowelsnl** [comment]
@ndellingwood -- is this an effect of upgrading to `CUDA-12`?

**Comment by crtrott** [comment]
OK I tracked down the root cause. Calling `cudaSetDevice` only affects the calling thread. Threads which get created later are back to a default environment, as if you never had called `cudaSetDevice`. This particular test actually does a threaded test. It does the following:

- initialize Kokkos (which calls `cudaSetDevice`)
- create instances from the default instance (i.e. create new cuda streams)
- create two std::threads
- dispatch kernel in each thread to a different instance

Asking for device 1, means the streams are created on device 1.
However the kernel dispatch on the new threads will try to dispatch to device 0, because nobody has called cudaSetDevice on them. That fails because the streams are only valid for device 1.

**Comment by PhilMiller** [comment]
Is this potentially going to blow up in Trilinos testing when we go to merge 4.0 into it? Does this need to be `Blocks Promotion`?

**Comment by ajpowelsnl** [comment]
> Is this potentially going to blow up in Trilinos testing when we go to merge 4.0 into it? Does this need to be `Blocks Promotion`?

I think so, based on what @ndellingwood said.

**Comment by PhilMiller** [comment]
Decision in today's meeting was that there are enough means for testing and users to work around this issue that it would not be considered blocking for 4.0

**Comment by ajpowelsnl** [comment]
Hi @ndellingwood , @masterleinad - is this issue resolved?  I've added to Kokkos Planning with the assumption that it's not.

**Comment by masterleinad** [comment]
> @ajpowelsnl Yes, the PR was merged #6299 late last week. The failing test here now passes.

@tcclevenger Did you actually test with device_id !=0?

**Comment by tcclevenger** [comment]
@masterleinad yes, I ran `KOKKOS_DEVICE_ID=1 ctest -R Kokkos_CoreUnitTests_Cuda1 -V` before and after and the error was resolved.

**Comment by masterleinad** [comment]
Fixed in #6299.
